lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -713,10 +743,39 @@ private void DoneAction()
         private void FailAction()
         {
             ShutDownAllEvaluators();
-            var msg = string.Format(CultureInfo.InvariantCulture,
-                "{0} The system cannot be recovered after {1} retries. NumberofFailedMappers in the last try is {2}, master evaluator failed is {3}.",
-                FailActionPrefix, _numberOfRetries, _evaluatorManager.NumberofFailedMappers(), _evaluatorManager.IsMasterEvaluatorFailed());
-            Exceptions.Throw(new ApplicationException(msg), Logger);
+            
+            var failMessage = _cancelEvent != null
+                    ? string.Format(CultureInfo.InvariantCulture,
+                        "{0} Job cancelled at {1}. cancellation message: {2}",
+                        FailActionPrefix, _cancelEvent.Timestamp.ToString("u"), _cancelEvent.Message)
+                    : string.Format(CultureInfo.InvariantCulture,
+                        "{0} The system cannot be recovered after {1} retries. NumberofFailedMappers in the last try is {2}, master evaluator failed is {3}.",
+                        FailActionPrefix, _numberOfRetries, _evaluatorManager.NumberofFailedMappers(), _evaluatorManager.IsMasterEvaluatorFailed());
+
+            Dispose();
+            Exceptions.Throw(new ApplicationException(failMessage), Logger);
+        }
+
+        /// <summary>
+        /// Dispose resources
+        /// </summary>
+        private void Dispose()
+        {
+            lock (_disposableResources)
+            {
+                _disposableResources.ForEach(handle =>
+                {
+                    try
+                    {
+                        handle.Dispose();
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.Log(Level.Error, "Failed to dispose a resource: {0}", ex);
+                    }
+                });
+                _disposableResources.Clear();
+            }
```

Comment:
I don;'t think so, IMRU driver does not create instance of JobLIfecycleManager. it just gets it as dependency.